### PR TITLE
Add ECDSA to providers

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
Depends on PR #10631 (EC key management)

**Comment by slontis** [comment]
Hmm looks like I might need to apply similiar changes that were just added to 
https://github.com/openssl/openssl/pull/10920

**Comment by levitte** [comment]
This needs a rebase

**Comment by slontis** [comment]
Dependant on #11133

File: test/recipes/30-test_evp_data/evppkey.txt [hunk 1]
```diff
@@ -635,7 +635,7 @@ Result = KEYOP_INIT_ERROR
 Reason = invalid salt length
 
 
-Title = EC tests
+Title = EC DSA tests
```
Comment (InfoHunter) [inline test/recipes/30-test_evp_data/evppkey.txt]: Is it intended to have a space between EC and DSA?

File: crypto/ec/ecdsa_aid.c [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#define ASN1_SEQUENCE 0x30
+#define ASN1_OID 0x06
+#define OID_FIRST(a, b) a * 40 + b
+#define DER_840() 0x86, 0x48    /* DER encoding of number 840 is 2 bytes */
+#define DER_10045() 0xCE, 0x3D  /* DER encoding of number 10045 is 2 bytes */
+#define SHA1_SZ 7
+#define SHA2_SZ 8
+#define SHA3_SZ 9
+
+/* ecdsaWithSHA1 OID is of the form : (1 2 840 10045 4 1) */
```
Comment (levitte) [inline crypto/ec/ecdsa_aid.c]: I would like to see information sources, like in crypto/rsa/rsa_aid.c

File: test/recipes/30-test_evp_data/evppkey.txt (hunk 1 above)
Comment (slontis) [reply test/recipes/30-test_evp_data/evppkey.txt]: removed

File: crypto/ec/ecdsa_aid.c [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+#include <stdlib.h>
+
+#include <openssl/objects.h>
+#include "crypto/ec.h"
+
+#define ASN1_SEQUENCE 0x30
+#define ASN1_OID 0x06
+#define OID_FIRST(a, b) a * 40 + b
+#define DER_840() 0x86, 0x48    /* DER encoding of number 840 is 2 bytes */
```
Comment (InfoHunter) [inline crypto/ec/ecdsa_aid.c]: Hi @slontis , how did you get these numbers? By checking the OID in a particular certificate file for instance?

Comment (slontis) [reply crypto/ec/ecdsa_aid.c]: Multiple ways..
This is just the DER encoding of any value > 127 as described here..
https://www.sciencedirect.com/topics/computer-science/basic-encoding-rules
It is also listed in mutliple online OID repositories as well as openssl's table of OIDS.

Comment (InfoHunter) [reply crypto/ec/ecdsa_aid.c]: Great, thanks

**Comment by slontis** [comment]
rebased now that dependant PR is merged to master

**Comment by slontis** [comment]
ping

File: providers/fips/fipsprov.c [hunk 4]
```diff
@@ -804,6 +804,9 @@ static const OSSL_ALGORITHM fips_keyexch[] = {
 static const OSSL_ALGORITHM fips_signature[] = {
 #ifndef OPENSSL_NO_DSA
     { "DSA:dsaEncryption", "provider=fips,fips=yes", dsa_signature_functions },
+#endif
+#ifndef OPENSSL_NO_EC
+    { "ECDSA:id-ecPublicKey", "provider=fips,fips=yes", ecdsa_signature_functions },
```
Comment (levitte) [inline providers/fips/fipsprov.c]: Please remove `:id-ecPublicKey` (it should be removed from the "ECDH" line as well).

Comment (levitte) [reply providers/fips/fipsprov.c]: To compensate, please fix these lines:

https://github.com/openssl/openssl/blob/ca7f7b951825e23dddb798f6a61f50a04225d25a/providers/implementations/keymgmt/ec_kmgmt.c#L52-L55

Comment (slontis) [reply providers/fips/fipsprov.c]: Done

Comment (levitte) [reply providers/fips/fipsprov.c]: Muuuch better, thank you

**Comment by slontis** [comment]
(Thanks for reviewing on the weekend :))

**Comment by slontis** [comment]
Travis issue is a timeout.

**Comment by slontis** [comment]
trivial rebase..

**Comment by slontis** [comment]
Merged to master.

**Pull request closed without merging**